Jetpack Assistant: Add UI

robertf4 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
This PR adds the Jetpack Assistant. It is the content of [the original PR](https://github.com/Automattic/jetpack/pull/17691) rebased against the repo reorg in master and replaces that PR. 

Site type prompt:
<img width="1054" alt="image" src="https://user-images.githubusercontent.com/42627630/105090292-dc58c780-5a63-11eb-8eeb-2d146a50896f.png">
<img width="365" alt="image" src="https://user-images.githubusercontent.com/42627630/105090380-eb3f7a00-5a63-11eb-9f53-8a1eb0f5fd4a.png">

WooCommerce prompt:
<img width="1060" alt="image" src="https://user-images.githubusercontent.com/42627630/105090434-001c0d80-5a64-11eb-912f-7a3bcba0ec4e.png">
<img width="367" alt="image" src="https://user-images.githubusercontent.com/42627630/105090453-07dbb200-5a64-11eb-81ec-dd2afb7dc5e2.png">

Monitor prompt:
<img width="1064" alt="image" src="https://user-images.githubusercontent.com/42627630/105090486-13c77400-5a64-11eb-97f6-bbce6c9247bc.png">
<img width="365" alt="image" src="https://user-images.githubusercontent.com/42627630/105090511-1c1faf00-5a64-11eb-8705-d74b00077c64.png">

Related posts prompt:
<img width="1061" alt="image" src="https://user-images.githubusercontent.com/42627630/105090541-26da4400-5a64-11eb-986a-7e901f6efdaa.png">
<img width="366" alt="image" src="https://user-images.githubusercontent.com/42627630/105090569-2f327f00-5a64-11eb-9af2-3f04350397c9.png">

Creative Mail prompt:
<img width="1053" alt="image" src="https://user-images.githubusercontent.com/42627630/105090607-3eb1c800-5a64-11eb-86be-c5b8fad37a6a.png">
<img width="366" alt="image" src="https://user-images.githubusercontent.com/42627630/105090633-45403f80-5a64-11eb-9369-7e65b1d87a3c.png">

Site accelerator prompt:
<img width="1055" alt="image" src="https://user-images.githubusercontent.com/42627630/105090682-55581f00-5a64-11eb-9a8e-4b2d7470f7e3.png">
<img width="362" alt="image" src="https://user-images.githubusercontent.com/42627630/105090699-5e48f080-5a64-11eb-8f31-dd49db9f5ac5.png">

Summary screen with no recommendations enabled/installed:
<img width="1056" alt="image" src="https://user-images.githubusercontent.com/42627630/105091570-a6b4de00-5a65-11eb-8e05-f599eaac3caf.png">
<img width="399" alt="image" src="https://user-images.githubusercontent.com/42627630/105092212-95200600-5a66-11eb-9b0c-4115e267bbd8.png">

Summary screen with all recommendations enabled/installed:
<img width="1060" alt="image" src="https://user-images.githubusercontent.com/42627630/105091992-496d5c80-5a66-11eb-8817-8d313bfeacef.png">
<img width="391" alt="image" src="https://user-images.githubusercontent.com/42627630/105091969-3fe3f480-5a66-11eb-8a8c-710bf6413c60.png">

Backup Daily sidebar upsell:
<img width="425" alt="image" src="https://user-images.githubusercontent.com/42627630/105092290-a9fc9980-5a66-11eb-99c4-c31966aa3db8.png">

Fallback sidebar upsell (shown if disabled in wpcom or wpcom can't be reached):
<img width="424" alt="image" src="https://user-images.githubusercontent.com/42627630/105093725-cf8aa280-5a68-11eb-96e3-78070a8ea662.png">

Enable one-click restores sidebar prompt:
<img width="423" alt="image" src="https://user-images.githubusercontent.com/42627630/105094196-88e97800-5a69-11eb-92d5-6c9983b11c58.png">

Manage your security sidebar prompt:
<img width="387" alt="image" src="https://user-images.githubusercontent.com/42627630/105094968-a4a14e00-5a6a-11eb-8c1d-9600e0f4418a.png">

Apps sidebar recommendation:
<img width="434" alt="image" src="https://user-images.githubusercontent.com/42627630/105093868-019c0480-5a69-11eb-93ee-f1be33ee7000.png">

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
MT: p1HpG7-asW-p2
Design: p6TEKc-46C-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

Initial prompts flow:
1. Start with a free site and navigate to `/wp-admin/admin.php?page=jetpack#/recommendations`.
2. Verify that you are redirected to `/wp-admin/admin.php?page=jetpack#/recommendations/site-type`.
3. Ensure that you have not selected "Store" and click "Continue".
4. Verify that you navigate to `#/recommendations/monitor`.
5. Click back in your browser and select "Store", then click "Continue".
6. Verify that you navigate to `#/recommendations/woocommerce`.
7. Navigate to `/wp-admin/index.php` and then back to `/wp-admin/admin.php?page=jetpack#/recommendations`.
8. Verify that you are redirected to `#/recommendations/woocommerce`.
9. Choose "Decide later for each of the prompts and verify that you are taken successively through `#/recommendations/monitor`, `#/recommendations/related-posts`, `#/recommendations/creative-mail`, and `#/recommendations/site-accelerator`.
10. Verify that the summary screen has no recommendations enabled.
11. Now return to `#/recommendations/monitor` and go back through the flow, selecting some recommendations and skipping others. Verify that the summary screen reflects your choices.
12. Enable the remaining recommendations on the summary screen and verify that it updates to reflect these changes.
13. Check the various "Learn more" and "Settings" buttons for each feature and make sure they are appropriate for that feature.
14. Check that the "View all Jetpack features' link works.

Summary upsell flows:
1. On a free site visit `/wp-admin/admin.php?page=jetpack#/recommendations/summary`.
2. Verify that an upsell prompt for Backup Daily shows.
3. Sandbox your site and in wpcom edit `/wp-content/rest-api-plugins/endpoints/jetpack-recommendations.php` so that the `WP_REST_Response` in `get_upsell()` returns `'hide_upsell' => true`.
4. Reload `#/recommendations/summary` and verify that the fallback sidebar upsell is displayed (no prices or product names should show).
5. Upgrade the site to Backup Daily.
6. Return to `#/recommendations/summary`. If the app sidebar prompt is displaying, wait a few minutes for Rewind to update, then reload the page.
7. Verify that the one-click restores prompt is displaying.
8. Use the prompt to add your credentials and then return to `#/recommendations/summary`. Verify that the "Manage your security" prompt displays in the sidebar and that the link works.
9. Start another free site and upgrade it to a paid plan that doesn't include backup or scan (e.g. anti-spam).
10. Navigate back to `#/recommendations/summary` and verify that the app sidebar prompt shows and that the app badges work.

Jetpack plugin upgrade flow:
1. Start with a site on an old version of Jetpack that includes the Setup Wizard (9.3 meets this criteria).
2. Use the Code Snippets plugin and add the following snippet to enable the Setup Wizard:
```
add_filter( 'jetpack_show_setup_wizard', '__return_true' );
add_filter( 'jetpack_pre_connection_prompt_helpers', '__return_true' );
```
3. Navigate to `/wp-admin/admin.php?page=jetpack#/setup/features` and click "I'm done for now" to complete the Setup Wizard.
4. On that site upgrade Jetpack so that it includes the Recommendations.
5. Verify that the Recommendations are not available from the dashboard, and that `#/recommendations` redirects to `#/dashboard`
6. Perform the above steps again but this time do not complete the Setup Wizard and verify that the Recommendations are available.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
Recommendations: add this new wizard to help you get started with recommended features of Jetpack.

jeherve commented on projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx:
```diff
@@ -0,0 +1,86 @@
+const MobileApp = () => {
+	const features = [
+		__( 'Refined post and page editor', 'jetpack' ),
+		__( 'Manage multiple sites from one dash', 'jetpack' ),
```
Should it be "dashboard"?

```suggestion
		__( 'Manage multiple sites from one dashboard', 'jetpack' ),
```

jeherve commented on projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/index.jsx:
```diff
@@ -0,0 +1,86 @@
+
+const MobileApp = () => {
+	const features = [
+		__( 'Refined post and page editor', 'jetpack' ),
+		__( 'Manage multiple sites from one dash', 'jetpack' ),
+		__( 'Multi-site plugin management', 'jetpack' ),
+		__( 'Free stock photo library', 'jetpack' ),
+		__( 'Update your site from any device', 'jetpack' ),
```
Should we include "Refined post and page editor" here, now that the editor in wpcom is an iFramed version of the wp-admin editor? Same with "Free stock photo library", now that the feature is available in wp-admin?

jeffgolenski replied:
hey hey @jeherve I think we should still include those list items. Even if they're available in wp-admin, many people probably don't know about them, and they're still good marketing tidbits for the apps I think.

jeherve commented on projects/plugins/jetpack/_inc/client/recommendations/sidebar/one-click-restores/index.jsx:
```diff
@@ -0,0 +1,88 @@
+						{ sprintf(
+							/* translators: placeholder is the name of a backups plan: Daily Backups or Real-time Backups */
+							__(
+								'Get the most out of your %s. One-click restores ensure we’ll be able to easily restore your site, if anything goes wrong.',
```
Should we say "you", since that's something you do on your own?

```suggestion
								'Get the most out of your %s. One-click restores ensure you’ll be able to easily restore your site, if anything goes wrong.',
```

jeherve reviewed (changes requested):
I gave this a try, and would have some questions.

********

The Progress status for the install buttons seems off a bit:

![image](https://user-images.githubusercontent.com/426388/105156209-cd2d5480-5b0b-11eb-8194-ac348b529063.png)

*********

When I tried to install Creative Contact, the installation failed silently (the button went back to "Install"), and here is the response I had in my console:

`Unable to install creative-mail-by-constant-contact: folder_exists`

The plugin had indeed been installed, but the installation had failed. It would be nice to surface any errors in the UI.

**********

When I tried to install Creative Contact again, the installation worked, but there was no indication of it in the UI. The line with the prompt to install disappeared, but that was the only hint that something had happened. I think it would be nice to show a notice with a link to the plugin wizard maybe?

Update: I see now that a line was added to the "enabled" section. I had missed that, I didn't expect that.

**********

I think the app store buttons could do with some extra padding (this was in Chrome)

![image](https://user-images.githubusercontent.com/426388/105157040-dec32c00-5b0c-11eb-9483-3e7511b6a790.png)

**********

Maybe we should stay focused on "Dashboard" when in recommendations?

![image](https://user-images.githubusercontent.com/426388/105168564-9579d900-5b1a-11eb-83bb-df5f7aeda347.png)

robertf4 commented on projects/plugins/jetpack/_inc/client/recommendations/feature-summary/index.jsx:
```diff
@@ -0,0 +1,111 @@
+			) }
+			<div className="jp-recommendations-feature-summary__display-name">{ displayName }</div>
+			<div className="jp-recommendations-feature-summary__actions">
+				<ExternalLink href={ learnMoreLink } onClick={ onLearnMoreClick }>
+					{ __( 'Learn more', 'jetpack' ) }
+				</ExternalLink>
```
That is intentional as it matches the Figma. `ExternalLink` only displays the icon when you set `icon=true`.

robertf4 commented on projects/plugins/jetpack/_inc/client/components/number-format/index.js:
```diff
@@ -33,7 +33,7 @@ export const cleanLocale = locale => {
  * @returns {string} Formatted user locale (e.g. `en-US` or `fr-FR`).
  */
-const getUserLocale = () => {
+export const getUserLocale = () => {
```
Good idea, I moved it to a `lib` folder.

robertf4 commented on projects/plugins/jetpack/package.json:
```diff
@@ -66,7 +66,9 @@
-		"@automattic/calypso-color-schemes": "1.0.0",
+		"@automattic/calypso-color-schemes": "2.1.1",
+		"@automattic/components": "1.0.0-alpha.3",
+		"@automattic/custom-colors-loader": "automattic/custom-colors-loader",
```
I am not sure how it ended up in here, but I don't think it is needed. I have removed it.

robertf4 commented on projects/plugins/jetpack/package.json:
```diff
@@ -66,7 +66,9 @@
 	},
 	"dependencies": {
 		"@automattic/calypso-build": "6.1.0",
-		"@automattic/calypso-color-schemes": "1.0.0",
+		"@automattic/calypso-color-schemes": "2.1.1",
```
I don't think so. I updated this because `@automattic/components` depends on having >=2.0. I can't find any usage of the package in Jetpack, and the changes were all renaming or removing variable names so I think the SCSS compiler would be failing if any of them was being used.

So the only change possible would be a change in a color somewhere, but looking around the plugin I have not noticed anything obvious.

robertf4 commented on projects/plugins/jetpack/_inc/client/state/products/reducer.js:
```diff
@@ -60,5 +60,5 @@ export function isFetchingProducts( state ) {
  */
 export function getProducts( state ) {
-	return state.jetpack.products.items;
+	return get( state.jetpack, [ 'products', 'items' ], {} );
```
At one point I was getting a null exception from this when it was accessed before it was initialized, but it's actually not being used anymore since I made a dedicated endpoint to get the upsell info. So I've reverted this change.

robertf4 commented on projects/plugins/jetpack/_inc/client/recommendations/sidebar/mobile-app/apps-badge.jsx:
```diff
@@ -0,0 +1,168 @@
+		defaultSrc: imagePath + '/get-apps-google-play.png',
+		src:
+			'https://play.google.com/intl/en_us/badges/images/generic/{localeSlug}_badge_web_generic.png',
+		tracksEvent: 'calypso_app_download_android_click',
```
Here I've just taken the component from Calypso and modified it slightly to work within Jetpack, so this is just me taking the behavior there.

In the event that loading the badge fails this component will default to using the local badge. For both iOS and Android the locale is in the URL. So if either the URL changes or if the language is not supported it will end up using the default English ones that are available locally.

robertf4 commented:
> The plugin had indeed been installed, but the installation had failed. It would be nice to surface any errors in the UI.

@jeffgolenski is currently mocking that up. Once I have the mockup I'll decide if it should be done in this PR or a subsequent one.

robertf4 commented:
> The Progress status for the install buttons seems off a bit

I've updated that so that it has a white color and has a better contrast against the button color.

> I think the app store buttons could do with some extra padding

I've added 32px padding above the app store buttons.

> Maybe we should stay focused on "Dashboard" when in recommendations?

This has been fixed.

jeffgolenski commented:
Posting the plugin install error state concept here for visibility:

If a plugin can't / won't install, we'll throw an error and provide a "manual install" option. Clicking / tapping that will take the customer directly to the plugin page within the plugin repo

![Uploading Screen Shot 2021-01-21 at 12.58.34 PM.png…]()

<img width="1678" alt="Screen Shot 2021-01-21 at 1 03 14 PM" src="https://user-images.githubusercontent.com/214813/105394126-3041e880-5beb-11eb-8ebc-466feceef04e.png">

jeherve reviewed (changes requested):
Thanks for the updates! This seems to test well for me at the moment. I noticed 2 things:

Looks like the linting check fails at the moment, for a small thing:
https://github.com/Automattic/jetpack/pull/18437/checks?check_run_id=1743317466

That's going to block the merge; do you think you could fix that?

*********

I noticed that the price of the backup product that was offered to me was different from the price I got when I clicked through:

![image](https://user-images.githubusercontent.com/426388/105397040-db20c980-5c20-11eb-902f-58e9b4be7009.png)
![image](https://user-images.githubusercontent.com/426388/105397063-e2e06e00-5c20-11eb-8496-9f92021245fc.png)

It would be nice to fix that.

jeffgolenski commented:
**Design Feedback w/ Video!**

Thanks for your hard work @robertf4. It's looking outstanding, I've just a few small details to clean up. I've documented each in video below:

**Large screens:**
video -> https://cloudup.com/cdnYwyeMrG7

items:
00:12 - option name php errors
00:28 - need to suppress any JITM messages on the "recommendations" page
00:40 - show recommendation screen on wp-admin dashboard
1:00 - increase description font size to 16px
1:55 - add more padding between right side of content area and image
3:08 - vertical alignment between star icon and "Recommended premium product" headline
3:25 - Recommended premium product card: add a couple px of space between "learn more" and new window icon
3:35 - make the "i" icon a lighter gray. "Gray 40" `#787C82`
3:44 - less space between "View all jetpack features" and new window icon. Let's make all of these consistent throughout entire experience for all "learn more" links on every screen.
3:54 - vertical gray border / divider is darker than the rest
4:04 - small white gap on the right side of the page. -1px right margin cleans it up, but there's probably a more semantic way to fix!
5:00 - "learn more" should open in new tab / window (I know I hate that idea too, but we don't have a clear flow to bring them back from jetpack.com!)
**5:00—5:30 - WooCommerce disappeared from the enabled list when I navigated out of wp-admin and back!**
6:15 - Namining inconsistency (my fault). Monitor is called "Downtime monitoring" in settings. We should make it all consistent.



**Mobile:** 

video -> https://cloudup.com/chfGAOTB6xm

items:
00:20 - padding inconsistency with top and bottom padding of checkboxes & labels
00:35 - bump up size of desc text as mentioned above!
1:00 - line height inconsistency in description text when "learn more" is present
1:35 - alignment issue of settings buttons on final page
1:50 - Originally recommend by @keoshi in a design review, I think we should make the feature titles become the "learn more" link, and actually remove the "learn more" links in our lists to free up space. See video for details
2:19 - reduce padding around "view all jetpack features" box to align with everything else around it

Thanks!

robertf4 commented:
> I noticed that the price of the backup product that was offered to me was different from the price I got when I clicked through

@jeherve I think this is due to a mismatch between the locale being used on the Jetpack site and on the wpcom user account. 

The call to retrieve the upsell price sends the locale from the Jetpack site [here](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php#L824), so the price displayed in the plugin should be based on the locale of the site. The clickthrough goes to cloud.jetpack.com, and there it will use the locale of the wpcom user because that is Calypso.

So I see 3 options here:
1. When retrieving the upsell price use the locale of the wpcom user account. This could result in prices/copy in the upsell card that do not match the language of the rest of the Jetpack plugin.
2. Use the Jetpack site's locale on cloud.jetpack.com. This could result in inconsistency if the user returns to cloud.jetpack.com through a different flow.
3. Leave it as-is, where there will be inconsistency if the Jetpack site and wpcom account have different locales set.

All of these are doable, but I'm not sure which one will create consistency for the greatest number of users. What are your thoughts?

jeherve commented:
> When retrieving the upsell price use the locale of the wpcom user account. This could result in prices/copy in the upsell card that do not match the language of the rest of the Jetpack plugin.

I would personally vote for this solution, but I'll be happy to let others weigh in with their decision on this.

robertf4 commented:
@jeffgolenski thank you for the review! I was able to address each point except for these ones.

> 00:28 - need to suppress any JITM messages on the "recommendations" page

Suppression of JITMs isn't done in the client, they should just not be configured to show on the recommendations page. I'm not sure how that one is showing in your video because they show based on the URL, and since the recommendations page has its own URL and is new there should be no JITMs for it. For example that particular JITM is set to only show at the root, dashboard, and media URLs. I haven't been able to get it to show on the recommendations page.

> 00:40 - show recommendation screen on wp-admin dashboard

This is coming in this PR: https://github.com/Automattic/jetpack/pull/18359

> WooCommerce disappeared from the enabled list when I navigated out of wp-admin and back!

I am not able to reproduce this. My only thought is that the plugins need a REST call display, and since jurassic ninja sites give 503 errors when anybody spins up a new site it is possible that you loaded it at just the right time for the call to the plugins endpoint to give a 503. Let me know if you are able to reproduce this.

> 1:00 - line height inconsistency in description text when "learn more" is present

The line height is set to 1.5 for everything here so I am not sure how to resolve this. Could you share the CSS that would resolve this?

robertf4 commented:
With regard to the plugin install error states I would like to move that to another PR as I think the recommendations can ship without it if needed, and I think finishing the banner is a higher priority.

jeffgolenski commented:
@robertf4 In regards to the line-height stuff, we can try to give the `a` links `display: inline-block;` which may clear them up.

jeherve reviewed (approved):
This tests well for me. I only seem to have troubles with the recommendation status.

> Jetpack plugin upgrade flow:
> [...]
> Verify that the Recommendations are not available from the dashboard, and that #/recommendations redirects to #/dashboard

This doesn't seem to work for me; the recommandations tab is always available to me:

- Even upgrading from a site where I had completed the wizard
- Even when the site is in Offline mode #18527  
- When navigating away from the React dashboard and then coming back to it, the recommendations are at the starting step again.

I think this can be solved in a follow-up PR though.

jeherve reviewed (changes requested):
I see that we have some linting errors that will need to be looked at before we can merge:
https://github.com/Automattic/jetpack/pull/18437/checks?check_run_id=1752060189
https://github.com/Automattic/jetpack/pull/18437/checks?check_run_id=1752061977

jeherve reviewed (approved):
Merging this now, so we can include it in 9.4.

jeherve commented:
> This doesn't seem to work for me; the recommandations tab is always available to me:
>
> Even upgrading from a site where I had completed the wizard
> ~~Even when the site is in Offline mode #18527~~
> When navigating away from the React dashboard and then coming back to it, the recommendations are at the starting step again.

Would you like me to create a follow-up issue to keep track of this, or is this already on your radar?

jeherve commented:
r220018-wpcom
